src/coap_blockwise.c

```diff
@@ -225,23 +251,32 @@ static void on_block_rcvd(struct golioth_client *client,
     ctx->is_last = is_last;
     memcpy(ctx->block_buffer, payload, payload_size);
     ctx->block_size = payload_size;
-    ctx->status = response->status;
+    ctx->response_status = response->status;
+
+    golioth_sys_sem_give(ctx->sem);
 }
 
 // Function to download a single block
 static enum golioth_status download_single_block(struct golioth_client *client,
                                                  struct blockwise_transfer *ctx)
 {
-    return golioth_coap_client_get_block(client,
-                                         ctx->path_prefix,
-                                         ctx->path,
-                                         ctx->content_type,
-                                         ctx->offset,
-                                         ctx->block_size,
-                                         on_block_rcvd,
-                                         ctx,
-                                         true,
-                                         GOLIOTH_SYS_WAIT_FOREVER);
+    enum golioth_status err = golioth_coap_client_get_block(client,
+                                                            ctx->path_prefix,
+                                                            ctx->path,
+                                                            ctx->content_type,
+                                                            ctx->offset,
+                                                            ctx->block_size,
+                                                            on_block_rcvd,
+                                                            ctx,
+                                                            false,
+                                                            GOLIOTH_SYS_WAIT_FOREVER);
+    if (GOLIOTH_OK == err)
+    {
+        golioth_sys_sem_take(ctx->sem, GOLIOTH_SYS_WAIT_FOREVER);
```

Comment:
I'm having a bit of trouble following how this change works. Will block downloads always block (prevent client from doing anything else until another chunk is received) or will there be an async option as well?

Reply:
Ah, I see this is going to be run on the [fw_update thread](https://github.com/golioth/golioth-firmware-sdk/blob/2e09cda7163d15212706ccfe24c1571ae7e5e396/src/fw_update.c#L170) so the blocking will happen there an not on the client thread.

Reply:
Yep, exactly right. Generally it will happen on the user thread, whatever that happens to be. I would like to get an asynchronous version later, where all operations happen on the client thread, but that's for the future.